graylog2-rest-client/src/test/java/org/graylog2/restclient/lib/AsyncByteBufferInputStreamTest.java

```diff
@@ -0,0 +1,167 @@
+
+        stream.putBuffer(ByteBuffer.wrap("123".getBytes()));
+        stream.putBuffer(ByteBuffer.wrap("456".getBytes()));
+
+        // we should be able to read 6 bytes in a row
+        byte[] bytes = new byte[6];
+        final int read = stream.read(bytes);
+        Assert.assertEquals(6, read);
+        Assert.assertArrayEquals("123456".getBytes(), bytes);
+        stream.setDone(true);
+
+        final int nextRead = stream.read();
+        Assert.assertEquals(-1, nextRead);
+        stream.close();
+    }
+
+    @Test
+    public void testDoneWithMoreData() throws InterruptedException, IOException {
+        final AsyncByteBufferInputStream stream = new AsyncByteBufferInputStream();
+
+        stream.putBuffer(ByteBuffer.wrap("123".getBytes()));
+        stream.putBuffer(ByteBuffer.wrap("456".getBytes()));
+
+        // we should be able to read 6 bytes in a row
+        byte[] bytes = new byte[6];
+        final int read = stream.read(bytes);
+        Assert.assertEquals(6, read);
+        Assert.assertArrayEquals("123456".getBytes(), bytes);
+
+        stream.putBuffer(ByteBuffer.wrap("789".getBytes()));
+        stream.setDone(true);
+
+        byte[] finalBytes = new byte[3];
+        final int nextRead = stream.read(finalBytes);
+        Assert.assertEquals(3, nextRead);
+        Assert.assertArrayEquals("789".getBytes(), finalBytes);
+
+        final int eos = stream.read();
+        Assert.assertEquals(-1, eos);
+
+        stream.close();
+    }
+
+    @Test
+    public void testAsync() throws InterruptedException {
+        final AsyncByteBufferInputStream stream = new AsyncByteBufferInputStream();
+
+        final Thread writer = new Thread() {
+            @Override
+            public void run() {
+                for (int i = 0; i < 10; i++) {
+                    stream.putBuffer(ByteBuffer.wrap("12345".getBytes()));
+                    stream.putBuffer(ByteBuffer.wrap("6\n".getBytes()));
+                    System.out.println("Wrote buffers step " + i);
+                    sleepUninterruptibly(500, TimeUnit.MILLISECONDS);
+                }
+                stream.setDone(true);
+
+            }
+        };
+        final Thread reader = new Thread() {
+            @Override
+            public void run() {
+                int singleByte = -1;
+                int count = 0;
+                do {
+                    try {
+                        singleByte = stream.read();
+                        count++;
+                    } catch (IOException e) {
+
+                    }
+                    if (count % 5 == 0) {
+                        System.out.println("read " + count + " bytes");
+                    }
+                } while (singleByte != -1);
+            }
+        };
+        reader.start();
+        writer.start();
+        reader.join();
+        writer.join();
+    }
+
+    @Test
+    public void testAsyncException() throws InterruptedException {
```

Comment:
Got it. I guess I would've used the [`expected`](http://junit.sourceforge.net/javadoc/org/junit/Test.html#expected()) attribute in the `@Test` annotation and not catch it in the reader thread to make that more obvious for inattentive people like myself.